scripts/release.py

```diff
@@ -0,0 +1,152 @@
+        )
+        logger.debug("script work dir: {}".format(script_work_dir))
+        for script in scripts:
+            try:
+                script_path = "{script_work_dir}/{script}".format(
+                    script_work_dir=script_work_dir,
+                    script=script
+                )
+                if os.path.exists(script_path):
+                    logger.debug("Executing script: {}".format(script_path))
+                    proc = subprocess.Popen(
+                        script_path,
+                        env=sub_env,
+                        cwd=script_work_dir
+                    )
+                    proc.communicate()
+                    if proc.returncode != 0:
+                        logger.error(
+                            "Script {s} failed. (Return Code: {rc})".format(
+                                s=script_path, rc=proc.returncode
+                            ))
+                        click.get_current_context().exit(-1)
+                else:
+                    logger.debug("Script not found: {}".format(script_path))
+                    if script.endswith("/run"):
+                        logger.error(
+                            "Failed to generate relese notes:"
+                            "run hook not found "
+                            "https://rpc-openstack.atlassian.net/wiki/"
+                            "spaces/RE/pages/19005457/RE+for+Projects"
+                        )
+                        click.get_current_context().exit(-1)
+            except Exception as e:
+                logger.error(traceback.format_exc())
+                logger.error("Failed to generate release notes: {}".format(e))
+                click.get_current_context().exit(-1)
+        if not os.path.exists(dst_file):
+            logger.error("Scripts did not generate release notes or did not"
+                         " place them in"
+                         " $WORKSPACE/{df}".format(df=dst_file))
+            click.get_current_context().exit(-1)
+    elif rnfile:
+        # Release notes are already in a file, just copy them to where
+        # they need to be
+        logger.debug("Reading release notes from file: {}".format(rnfile))
+        return_code = os.system("cp {src_file} {dst_file}".format(
+            src_file=pipes.quote(rnfile),
+            dst_file=dst_file)
+        )
+        if return_code != 0:
+            logger.error(
+                "Failed to read release notes file: {}".format(rnfile))
+            click.get_current_context().exit(-1)
+    elif text:
+        logger.debug("Release notes supplied inline: {}".format(text))
+        with open(dst_file, "w") as out:
+            out.write(text)
+    else:
+        logger.error("One of script, text or file must be specified. "
+                     "No release notes generated")
+        click.get_current_context().exit(-1)
+    with open(dst_file, "r") as notes:
+        ctx_obj.release_notes = notes.read()
```

Comment:
I would leave this out. The only place it's used, `create_release`, can read the file itself. This mechanism makes `create_release` harder to read.

Reply:
Generate release notes exposes all the mechanisms for obtaining release notes, from a script, text or file. It made sense to me to keep them all as options within one command.

Reply:
I'm don't follow what you are saying here. I'm not suggesting removing support for script/file/text. I'm referring to the use of `ctx_obj` which doesn't add any functionality in this case given that `create_release` will read from a file.

Reply:
The functionality added is that all the release notes sources are options to the same command, rather than two on generate_release notes and one on create release. For a user it probably makes sense to have all the options for adding release notes in one place?

Reply:
I think you are misunderstanding me, I'm referring to the following lines and the associated code in other functions:
```python
    with open(dst_file, "r") as notes:
        ctx_obj.release_notes = notes.read()
```
You've coupled `generate_release_notes` with `create_release`. `generate_release_notes` hardcodes the filename where it writes the release notes but also adds it to `ctx_obj`. `create_release` checks `ctx_obj` but also takes a filename `bodyfile`, which is specified in `rpc_jobs/release.yml`, although this is not used.

I think it would be simpler if both `generate_release_notes` and `create_release` simply communicated via the file and so `generate_release_notes` should take the same `--bodyfile` argument instead of hardcoding the path.

Reply:
I agree that release.yml should not specify `--bodyfile` for create_release. I think that is leftover from before the context object was used. `create_release` could be used independently of generate_release_notes by using the bodyfile argument you mention. 

The path is hard coded because its part of the specification in RE for projects. Its very simple specify a path relative to the workspace where a script must write release notes to. I initially used stdout, but that made implementation more complicated as you have to be careful about unintended leaks to stdout.

Maybe it would be better to keep all the hard coded paths in the job, rather than in the python? As the pre/run/post scripts are currently.

Reply:
I think you've agreed with me while having a completely different opinion.
- the context object is a hinderance not a help
- a specification that requires a script to write to a certain location is less simple than one that requires a script to write to a configurable location. If you can guarantee that location will never change, then hardcode it but I don't see how it's ever possible to make that guarantee.
- Yes, lets have one place where this config is set.